lib_xud/src/user/client/XUD_EpFunctions.c

```diff
@@ -136,36 +142,139 @@ XUD_Result_t XUD_GetBuffer_Finish(chanend c, XUD_ep e, unsigned *datalength)
     /* Input tail length (bytes) */
     asm volatile("int %0, res[%1]" : "=r"(lengthTail) : "r"(c));
 
+    unsigned frame;
+    asm volatile("int %0, res[%1]" : "=r"(frame) : "r"(c));
+
     /* Bits to bytes */
     lengthTail >>= 3;
 
     /* Words to bytes */
     length <<= 2;
 
     /* -2 length correction for CRC */
-    *datalength = length + lengthTail - 2;
+    recv_length = length + lengthTail - 2;
+
+#if (XUD_USB_ISO_MAX_TXNS_PER_MICROFRAME > 1)
+    unsigned ep_marked_ready = 0;
+    if(!ep->out_err_flag && ep->actualPid == USB_PIDn_MDATA)
+    {
+        // We expect more data
+        ep->buffer += recv_length;
+        /* Mark EP as ready */
+        unsigned * array_ptr = (unsigned *)ep->array_ptr;
+        *array_ptr = (unsigned) ep;
+        ep_marked_ready = 1;
+    }
+    unsigned got_sof = (ep->saved_frame != frame) ? 1 : 0;
+    ep->saved_frame = frame;
+#endif
 
+#if (XUD_USB_ISO_MAX_TXNS_PER_MICROFRAME == 1)
     /* Load received PID */
     unsigned receivedPid = ep->actualPid;
 
     /* Check received PID vs expected PID */
     if(receivedPid != ep->pid)
     {
         *datalength = 0; /* Extra safety measure */
+        XUD_Result_t res = XUD_GetBuffer_Start(ep, ep->buffer);
+        if(res == XUD_RES_UPDATE) return res;
         return XUD_RES_ERR;
     }
+    else
+    {
+        *datalength = recv_length;
+    }
+#endif
 
     /* ISO == 0 */
     if(ep->epType != XUD_EPTYPE_ISO)
     {
+#if (XUD_USB_ISO_MAX_TXNS_PER_MICROFRAME > 1)
+        /* Load received PID */
+        unsigned receivedPid = ep->actualPid;
+
+        /* Check received PID vs expected PID */
+        if(receivedPid != ep->pid)
+        {
+            *datalength = 0; /* Extra safety measure */
+            XUD_Result_t res = XUD_GetBuffer_Start(ep, ep->buffer);
+            if(res == XUD_RES_UPDATE) return res;
+            return XUD_RES_ERR;
+        }
+        else
+        {
+            *datalength = recv_length;
+        }
+#endif
 #ifdef __XS2A__
         ep->pid ^= 0x8;
 #else
         ep->pid ^= 0x88;
 #endif
     }
-
+#if (XUD_USB_ISO_MAX_TXNS_PER_MICROFRAME > 1)
+    else
+    {
+        /// Limitation: We only support a max of 2 transactions per OUT transfer. ep->current_transaction can only be 0 or 1
+        unsigned current_transaction = ep->current_transaction;
+        unsigned error = ep->out_err_flag; // If we're already in error from last transaction
+        if(current_transaction == 0)
+        {
+            // 1st transaction can only be DATA0 or MDATA and it should see a sof
+            if(!got_sof || (ep->actualPid == USB_PIDn_DATA1))
```

Comment:
error = !got_sof || ep->actualPid == USB_PIDn_DATA1) maybe?

Depends what you think is easier to read. Compiler output might be slightly different.

Reply:
done